scrapy_poet/injection.py

```diff
@@ -0,0 +1,323 @@
+import inspect
+import logging
+from typing import Dict, Callable, Any, List, Set
+
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+import andi
+from scrapy import Request, Spider
+from scrapy.crawler import Crawler
+from scrapy.http import Response
+from scrapy.settings import Settings
+from scrapy.statscollectors import StatsCollector
+from scrapy.utils.defer import maybeDeferred_coro
+from scrapy.utils.misc import load_object
+from scrapy_poet.injection_errors import (UndeclaredProvidedTypeError,
+                                          NonCallableProviderError,
+                                          InjectionError)
+from scrapy_poet.page_input_providers import PageObjectInputProvider, PROVIDERS
+from scrapy_poet.api import _CALLBACK_FOR_MARKER, DummyResponse
+from web_poet.pages import is_injectable
+
+
+logger = logging.getLogger(__name__)
+
+
+class Injector:
+    """
+    Keep all the logic required to do dependency injection in Scrapy callbacks.
+    Initializes the providers from the spider settings at initialization.
+    """
+    def __init__(self, crawler: Crawler):
+        self.crawler = crawler
+        self.spider = crawler.spider
+        self.load_providers()
+
+    def load_providers(self):
+        self.providers = [
+            cls(self.crawler)
+            for cls in load_provider_classes(self.spider.settings)
+        ]
```

Comment:
I wonder if that'd be better to follow "scrapy way" here, and configure providers similarly to middlewares / pipelines / extensions. In Scrapy they're configured with `{cls: priority}` dicts in the user code, and there is also a base setting as well. So, you'd have

`SCRAPY_POET_PROVIDERS_BASE = {ResponseDataProvider: 500}`,

and then users can do `SCRAPY_POET_PROVIDERS = {MyProvider: 100, 'path.to.MyPovider2': 200}`.

The code to handle it is implemented here: 

* https://github.com/scrapy/scrapy/blob/85604e1078b927c2a875040e29c58b4594c8d386/scrapy/core/spidermw.py#L31
* https://github.com/scrapy/scrapy/blob/85604e1078b927c2a875040e29c58b4594c8d386/scrapy/utils/conf.py#L15
* https://github.com/scrapy/scrapy/blob/85604e1078b927c2a875040e29c58b4594c8d386/scrapy/middleware.py#L29 (note that create_instance is used there).

TBH, I don't have strong reasons to do this change, besides consistency.

Reply:
Working on changing the settings, but one question, what do you think about the `from_crawler`? I can stick to the standard of using the `from_crawler` or just keeping as is now and receive always the `crawler` as `__init__` argument. I'm inclined to the latter because I see the `from_crawler` as boilerplate, but this would break somehow the standard. @kmike what do you think?

Reply:
Also another question regarding `SCRAPY_POET_PROVIDERS_BASE`. This works well for Scrapy as you can include the default in [default_settings.py](https://github.com/scrapy/scrapy/blob/5e9cc3298be4ca1146d68098899365b334706db8/scrapy/settings/default_settings.py), but I don't think we have a good way to access to that from `scrapy-poet`. So I'll have a local default for this case with the default provider.

Reply:
Implemented in https://github.com/scrapinghub/scrapy-poet/pull/41/commits/b176600581570bb9a6a5606b3d5cb7b206e88fa9

Finally decided not to follow `from_crawler` standard